### PR TITLE
Update supported versions: remove Python 3.7 and add Python 3.12

### PR DESCRIPTION
**Pull request by msabramo**
as described at https://devguide.python.org/versions/

Fixes: GH-971

There's probably more that needs to be done with GitHub Actions, `README.rst`, `changelog.rst`, but this is a start...

### CODE REVIEW

**Comment by BoboTiG** [comment]
You can grep the whole source code with "3.7" & "37" and adapt here and there.

**Comment by msabramo** [comment]
I seeked and destroyed I think the relevant cases of `3.7` and `37`. What's remaining seems okay to keep I guess:

```
(.venv)
abramowi at Marcs-MacBook-Pro-3 in ~/Code/OpenSource/watchdog (update-supported-versions)
$ ag '3\.7|37'
tests/test_observers_winapi.py
117:        File "C:\Python37-32\lib\threading.py", line 926, in _bootstrap_inner

changelog.rst
11:- Drop support for Python 3.7.
224:- Avoid deprecated ``PyEval_InitThreads`` on Python 3.7+ (`#746 <https://github.com/gorakhargosh/watchdog/pull/746>`__)
343:- [inotify] Fixed the way ``OSError`` are re-raised (`#377 <https://github.com/gorakhargosh/watchdog/issues/377>`__)
375:- [bsd] Improved the platform detection (`#378 <https://github.com/gorakhargosh/watchdog/pull/378>`__)
376:- [inotify] Fixed a crash when the root directory being watched by was deleted (`#374 <https://github.com/gorakhargosh/watchdog/pull/374>`__)
382:- [windows] Fixed the ``FILE_LIST_DIRECTORY`` constant (`#376 <https://github.com/gorakhargosh/watchdog/pull/376>`__)

src/watchdog_fsevents.c
724:// PyEval_InitThreads() does nothing as of Python 3.7 and is deprecated in 3.9.
```

**Comment by BoboTiG** [comment]
Maybe also tweak:
- https://github.com/gorakhargosh/watchdog/blob/6cfe9cc3e84b29264ae8c5fa1f1b3ce9c09acfbf/.github/workflows/build-and-publish.yml#L56
- https://github.com/gorakhargosh/watchdog/blob/6cfe9cc3e84b29264ae8c5fa1f1b3ce9c09acfbf/.github/workflows/tests.yml#L44

**Comment by msabramo** [comment]
Haha I was *just* going to say that I just remembered that `ag` ignores dotfiles and dot directories by default, so I think I missed the GitHub Actions stuff.

**Comment by BoboTiG** [comment]
I think you can add 3.12 to the CI + tox. It will fail for a time, but then it will run automagically when tox will support 3.12 (without having us to check for updates on their side).

File: changelog.rst [hunk 1]
```diff
@@ -8,6 +8,7 @@ Changelog
 
 2023-xx-xx • `full history <https://github.com/gorakhargosh/watchdog/compare/v3.0.0...HEAD>`__
 
+- Drop support for Python 3.7.
```
Comment (BoboTiG) [inline changelog.rst:11]:
```suggestion
- Drop support for Python 3.7.
- Add support for Python 3.12.
```

**Pull request merged**